mm/mm_heap/mm_free.c

```diff
@@ -78,34 +78,37 @@ static void add_delaylist(FAR struct mm_heap_s *heap, FAR void *mem)
  *
  ****************************************************************************/
 
-void mm_delayfree(FAR struct mm_heap_s *heap, FAR void *mem, bool delay)
+void mm_delayfree(FAR struct mm_heap_s *heap, FAR void *mem,
+                  bool delay, bool cleanup)
 {
   FAR struct mm_freenode_s *node;
   FAR struct mm_freenode_s *prev;
   FAR struct mm_freenode_s *next;
   size_t nodesize;
   size_t prevsize;
 
-  if (mm_lock(heap) < 0)
+  if (cleanup)
     {
-      /* Meet -ESRCH return, which means we are in situations
-       * during context switching(See mm_lock() & gettid()).
-       * Then add to the delay list.
-       */
-
-      add_delaylist(heap, mem);
-      return;
-    }
-
 #ifdef CONFIG_MM_FILL_ALLOCATIONS
-  memset(mem, MM_FREE_MAGIC, mm_malloc_size(heap, mem));
+      memset(mem, MM_FREE_MAGIC, mm_malloc_size(heap, mem));
 #endif
 
-  kasan_poison(mem, mm_malloc_size(heap, mem));
+      kasan_poison(mem, mm_malloc_size(heap, mem));
+    }
 
   if (delay)
     {
-      mm_unlock(heap);
+      add_delaylist(heap, mem);
+      return;
+    }
+
+  if (mm_lock(heap) < 0)
+    {
+      /* Meet -ESRCH return, which means we are in situations
+       * during context switching(See mm_lock() & gettid()).
+       * Then add to the delay list.
+       */
+
```

Comment:
I get what you are trying to do but the coding (to get the comments in place) looks poor. If (A) then if (B) do the same thing.